### DISCUSSION TITLE
Plans on the horizon

### DISCUSSION THREAD

**Discussion by jacobslusser** [sshnet/SSH.NET#1269] [General]
Hello @Rob-Hague (CC @WojciechNagorski).

As mentioned in my recent PR, I have a few focus areas that I thought I might start working on and wanted to give you a preview. In no particular order...

**Including `ILogger` support and more verbose tracing options**
This is initially what I came to the project source code looking for. Where I work, we make fairly extensive use of the SFTP features of SSH.NET (thank you) and occasionally connections fail and I would be interested in getting more information about the cause. So I'm looking to update some of the exceptions, add `ILogger` uses, or both.

**Performance improvements: heap allocations**
I've only just started doing some profiling but can see there are lots of areas we are allocating `byte[]` arrays that could benefit from using `ReadOnyMemory<byte>` or pooling. For example, `Message.GetPacket` popped up on a few of my early profiles that could benefit from some optimization.

**Performance improvements: async all the way down**
In the aforementioned profiling, I also spotted `SocketAbstractions.Send` and have a `SendAsync` version that I am testing and will eventually submit a PR for. Lots of other places that could benefit from `Async` version of methods, but as mentioned in our discussion [here](https://github.com/sshnet/SSH.NET/issues/1259#issuecomment-1844194006), some of which will need discussion with the group because they could make breaking changes in the API or require a lot of conditional compilation. 

**AOT compilation support**
I found that `Message.WriteBytes` is doing a lookup of `MessageAttribute` on every packet/message write and contributes a non-trivial amount to the overhead of each message. This got me thinking about not only caching/eliminating the lookup for this attribute, but for AOT down the road. Much longer term goal, but I ran the AOT analyzer to identify problem areas and it doesn't look too difficult.

**Documentation**
I would be happy to update some of the docs (or add them) and/or the wiki with some more information.

So that's what has popped into my head in the first few days of spending time with the code. I'm sure I will come across more areas that will look like fun problems to work on.

Some question that might help me and others:
- Is there a place where roadmap like work is tracked? I'm assuming from Rob's [Infrastructure work items issue](https://github.com/sshnet/SSH.NET/issues/1259), that using the Issues is the approach for documenting roadmap like work and discussing in.
- Do we like the wiki for documentation or do we want to create a `docs/` folder with markdown files? I tend to prefer the second option since it lives with the repo instead of with GitHub.
- What is our release cadence and methodology? I'm already seeing possible breaking changes to the API which we would only want to entertain in a major version bump, but not sure when or how we decide that.

**Comment by Rob-Hague** [comment]
Thanks for the write-up!

> Including ILogger support and more verbose tracing options

There is some diagnostics using (legacy) TraceSource (see the [wiki](https://github.com/sshnet/SSH.NET/wiki/Troubleshooting-SSH.NET#net-tracesource-and-tracelisteners) and #1263). Using [Microsoft.Extensions.Logging](https://www.nuget.org/packages/Microsoft.Extensions.Logging/) would be a good upgrade.

> Performance improvements: heap allocations

Agreed, sounds good.

> Performance improvements: async all the way down

A good thing that I have also been thinking about lately. We should eventually look to replace all the Begin/End ("APM") methods with truly async implementations (the current implementations mostly just dispatch onto the threadpool). For this we need to make `Session.MessageListener` async, and we need to create `Session.SendMessageAsync`. `Session.SendMessage` uses locking, so to make this asynchronous we should probably convert that lock into a `SemaphoreSlim` which has `WaitAsync` (the motivation behind #1265 🙂) or a type which wraps `SemaphoreSlim` and looks like a lock (I have a small prototype for that).

> AOT compilation support

See #1216 which fixes the warnings. Replacing the `MessageAttribute` system could be good for performance. In effect that is just specifying the message number, which could just be e.g. defined on an interface.

> Documentation

Great. The README could perhaps also get a refresh with more examples. e.g. by following the template of the Microsoft.Extensions.Logging package linked above.

> Is there a place where roadmap like work is tracked

No, but this discussion seems as good as any

> Do we like the wiki for documentation

I have no preference here (if someone is motivated to write documentation, then whichever format makes them the most motivated is good for me). I will defer to others on that one.

> What is our release cadence and methodology?

I don't know if there is one. For some background, the library has been nobly stewarded by approximately 1 maintainer for most of its lifetime, and has only recently started to get a bit more love. So things are not so well-defined in that sense. But from my short experience, breaking changes are reasonable particularly in the lower-level areas.

---

From my side, my current list of interests (I tend to jump between things - I know it's bad - I'm not reserving any for myself):

* ~~Investigate/stabilise random failing tests in CI~~ #1253 / #957
* ~~Add support for OpenSSH certificates~~ #1498
* ~~Send client KEXINIT first (see discussion in #972, I'm running a version of this in production)~~ #1274
* ~~Add support for aes-gcm cipher and etm hmacs~~ #1316 and #1369
* Async as discussed
* ~~Replace cipher logic with BCL implementations~~ #1373, #1458, #1546
* ~~Fix ShellStream and PipeStream~~ #1322 and #1399
* Null-annotate the library
* ~~Replace BigInteger with BCL version~~ #1469
* ~~Replace the dictionaries in `ConnectionInfo` with a type that guarantees order~~ #1611
* Relax the endpoint logic (remove DNS lookups?) in the forwarded ports/SOCKS implementations (something like #1234 #1072)
* ~~Merge (or at least add) the "Shared" tests project into the solution~~ #1467

**Reply by fmargueirat** [reply]
It would be great to see aes-gcm ciphers in main branch. I saw there are some PRs for it, but I was not able to get them to work, after downloading the code and compile it locally. I'm a fairly noob when it comes to GIT (I usually download approved/latest releases), so I am probably missing some steps.

**Reply by Rob-Hague** [reply]
It's possible those PRs simply don't build (I didn't check). The library these days does not require anything special to build it, but that might not have been true when they were last updated (quite a lot has changed regarding the build "infrastructure" since then)

**Reply by fmargueirat** [reply]
@Rob-Hague, I think the PRs are for older versions and the code on which they were forked eventually changed and that's why they are not building. I still think adding the support in the latest version, would be a good idea.

**Reply by scott-xu** [reply]
@fmargueirat aes-gcm ciphers are added in develop branch (.NET 6.0 onward only). See https://github.com/sshnet/SSH.NET/pull/1369.
With https://github.com/sshnet/SSH.NET/pull/1450, it expands support for all targeted frameworks.

**Comment by jacobslusser** [comment]
Hello friends. Life got busy but I would be interested in contributing back again. I don't think my goals have changed much, so I thought I would pick up where I left off. 

Question: how do we feel about simple cleanup PRs? Not like massive restructuring stuff, but just little things. For example, today I spotted a few places where we have `CancellationToken cancellationToken` but likely want it to be `CancellationToken cancellationToken = default`. Or something like `Session_ErrorOccured` corrected to `Session_ErrorOccurred`. Little work like this can sometimes help my brain get reacquainted with a codebase, but wanted to check that we are okay with those kinds of PRs.

**Reply by Rob-Hague** [reply]
In principle sure. I don't really enjoy changing public API, but internal stuff is no issue. Would generally prefer to review a smaller number of grouped cleanup changes than many tiny changes separated out.

If you want some suggestions for worthwhile low-impact changes:

1. Many of the tests in the unit test project are really difficult to work with: test classes with lengthy/strict mock setups which usually only test one or two things per class and are then duplicated many times, only changing very slightly. They tend to heavily tie-down the implementation so it can feel like the implementation of something exists once in the library and 10 times in the test project... It should be possible to de-duplicate and simplify a lot of the tests, which will help future development. If you are interested in doing this, ignore the SftpFileStream* and SftpFileReader* tests - I will tackle those.
2. In the integration test project, I think there are probably some tests which don't need to be integration tests and could just as well be unit tests. We have a Windows CI leg now but it is much slower than the Linux leg so trimming unnecessary fat is worthwhile. Example: https://github.com/sshnet/SSH.NET/blob/6bb397872b3a064109780f68735d5c4586e992fd/test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SftpClientTest.DeleteDirectory.cs#L52-L60
3. Along the same lines, I think there are probably some duplication in the integration tests, between the top-level directory and the tests in the "OldIntegrationTests" sub-directory. Merging these would be something to look at.
4. In the library code, a lot of the xmldoc is duplicated from the base type onto the implementing type, when it could just use `/// <inheritdoc />`. In general any documentation improvements are welcome (including the examples in docfx/examples.md).
